### ISSUE TITLE
Can't (directly) copy commands from detect-crop on Windows

### ISSUE THREAD

**Issue by forty2** [donmelton/video_transcoding#146]
First, a big "thank you" for the time and effort you've put into these scripts -- they've been producing great results for me ever since I found them.

One minor annoyance I've come across, though, is that when I'm using Windows to detect crop values, I can't directly copy and paste the preview commands out of the `detect-crop` output.

`detect-crop` produces commands that look like this:
```
mpv --no-audio --vf 'lavfi=[drawbox=10:0:704:576:invert:1]' Some\ File\ With\ Spaces.mkv
```
but `cmd.exe` doesn't support either single quotes or backslash escaping for filenames.  Would it be possible, at least on Windows, to produce commands like this instead:
```
mpv --no-audio --vf "lavfi=[drawbox=10:0:704:576:invert:1]" "Some File With Spaces.mkv"
```
?  It would save me a bunch of time, especially when I have a big batch of files to process, to not have to manually edit each command.

Thanks again!

**Comment by donmelton** [comment]
@forty2 You are very welcome, sir! I'm glad they're working for you.

And my apologies for the bug, but I can't easily fix that since Ruby doesn't provide a way for me to escape strings in a cross-platform way other than the `shellescape` mechanism designed for Bourne-compatible shells. It doesn't really work for `cmd.exe` on Windows.

I thought I had documented that in the "README" but apparently that's not the case. I can at least fix that. And I'm open to suggestions or patches to change the code itself.

Just out of curiosity, have you tried this in Windows PowerShell?

**Comment by forty2** [comment]
Well, that's a bummer, but thanks for the explanation.  I hadn't thought to try PowerShell; unfortunately it doesn't work either (the single quotes are fine, but it still can't figure out the backslashes), but that did prompt me to think of trying the bash that comes with Git.  And that worked just fine (as one might expect).

So, bottom line, I've got a useful workaround.  Thanks!  (A note in the README might be nice, though, if/when you get around to it.)

**Comment by donmelton** [comment]
@forty2 Whoops! It looks like I closed this last night when I thought I was only commenting on it. That's what I get for drinking my dinner. :) My sincere apologies for being so fat fingered.

And thanks for finding that incredibly simple workaround. I'll probably add that to the "README" as well.

But I hope @samhutchins can suggest a real fix for this since he's one of our resident Windows experts.

**Comment by donmelton** [comment]
@forty2 Also, thanks so trying PowerShell so quickly. I don't have access to my Windows computer where I am at the moment so testing is a bit tricky for me.

**Comment by samhutchins** [comment]
On Windows if you wrap the filename in double quotes it works, is that not the case on other platforms?

**Comment by samhutchins** [comment]
I've done a little bit of digging, and made a local change to demonstrate the idea:

```
diff --git a/bin/detect-crop b/bin/detect-crop
index 892ab8b..79a72d4 100755
--- a/bin/detect-crop
+++ b/bin/detect-crop
@@ -90,7 +90,7 @@ HERE
       directory = media.info[:directory]
       hb_crop = media.info[:autocrop]
       hb_crop = Crop.constrain(hb_crop, width, height) if @constrain
-      shell_path = arg.shellescape
+      shell_path = "\"#{arg}\""
 
       print_transcode = ->(crop) do
         puts
@@ -104,7 +104,7 @@ HERE
         puts
 
         if @player == :mpv
-          puts "mpv --no-audio --vf 'lavfi=[drawbox=#{Crop.drawbox_string(crop, width, height)}:invert:1]' #{shell_path}"
+          puts "mpv --no-audio --vf lavfi=[drawbox=#{Crop.drawbox_string(crop, width, height)}:invert:1] #{shell_path}"
           puts "mpv --no-audio --vf crop=#{Crop.player_string(crop, width, height)} #{shell_path}"
         else
           str = Crop.player_string(crop, width, height)
```

The result of this change: 
```
G:\Video\Unprocessed rips\Review\Howl's_Moving_Castle - done>detect-crop.bat "Howl's Moving Castle t04.mkv"
...

mpv --no-audio --vf lavfi=[drawbox=0:22:1920:1036:invert:1] "Howl's Moving Castle t04.mkv"
mpv --no-audio --vf crop=1920:1036:0:22 "Howl's Moving Castle t04.mkv"

transcode-video --crop 22:22:0:0 "Howl's Moving Castle t04.mkv"
```

This makes the command directly copy-able on Windows, but effectively reverts 96c490511c5f2fef445791c405acbe3c4333058c in the process. I dunno how to deal with Z shells without breaking Windows. Not to say it can't be done, just that I've never used (or really heard of) Z shells

I also don't know the effects of just wrapping the arg in `"` instead of `shellescape` is on Unix-like systems. I can't easily test today, the only Mac I have access to is in the office

**Comment by donmelton** [comment]
@samhutchins Thanks for doing all this detective work, testing and coming up with a patch! You are thorough, as usual.

And you're right about the change I made to fix `detect-crop` for Z shells. It made things worse for Windows. But that's a (slightly) separate issue that I need to address.

Your patch will likely work for the common case, even across multiple platforms. In fact, it looks like my original implementation before I realized that I had to use `shellescape`. But there's a big problem with it which also explains my decision to switch. And something you already suspected. It won't handle `"`s (quotation marks) and a few other symbols already embedded in the path string.

So, I went searching around teh interwebs this morning to see if anyone else has solved this problem. And I found a good news/bad news answer here:

http://stackoverflow.com/questions/5636601/is-there-an-equivalent-to-the-ruby-shellwords-module-for-the-windows-shell

Basically, I have to code a wrapper for `shellescape` that special cases for Windows. Which is something I was hoping to avoid since that would be the only platform-specific code in the project. And, BTW, the example code on that page likely has a bug in it having to do with those pesky embedded quotation marks.

Anyway, I'm away from my main development machine at the moment, but as soon as I get back I'll take a crack at coming up with something. And likely I'll ask you to help me test it. :)

Thanks again!

**Comment by donmelton** [comment]
@samhutchins I just notice a pointer to this code:

https://github.com/larskanis/shellwords

...at the bottom of that Stack Overflow article I linked to in the previous comment. Unfortunately, the `cmd.exe` version of `shellescape` is not implemented.

**Comment by donmelton** [comment]
@samhutchins After even more snooping around on the Web, it looks like this is the best way to detect whether we're running on Windows:

```
if Gem.win_platform?
```

...because the Gem module is included with Ruby version 1.9 and later. That way I don't need to add a dependency on the the OS module. See this page for discussion of that issue:

http://stackoverflow.com/questions/4871309/what-is-the-correct-way-to-detect-if-ruby-is-running-on-windows

But it occurs to me that it's not really Windows I should care about so much as whether the current shell is `cmd.exe`. After all, you can easily run Bash on Windows. And I have no idea how to detect the current shell. So I'm not even sure fixing this problem is even possible given that constraint.

Any ideas this new issue?

**Comment by samhutchins** [comment]
That's cool that it's possible to detect OS without another dependency. I'll have a think about how to detect CMD, nothing immediately springs to mind though

**Comment by donmelton** [comment]
@samhutchins OK, I think I've found a way to detect which shell is running. It's kind of a hack, but it should work. Can you try this on your Windows machine since I don't have access to one at the moment? From `cmd.exe` invoke `irb`, the interactive Ruby environment, and type this at the prompt:

```
RbConfig::CONFIG['SHELL']
```

On macOS that prints out:

```
=> "/bin/sh"
```

I would be curious to know what it does on Windows.

Thanks!

**Comment by samhutchins** [comment]
I found this: http://stackoverflow.com/questions/5859854/detect-if-bat-file-is-running-via-double-click-or-from-cmd-window

On windows all Gems are launched via a `.bat` file, which means the `cmdcmdline` environment variable is set. I just made a little test batch file that calls a test Ruby file that prints the `cmdcmdline` environment variable. 

Via CMD:
```
C:\Users\Sam>foo.bat
"C:\Windows\system32\cmd.exe"
```

Via Powershell:
```
PS C:\Users\Sam> .\foo.bat
C:\Windows\system32\cmd.exe /c ""C:\Users\Sam\foo.bat""
```

I don't have a bash on my PC, so I can't easily test that right now. If someone is using the Bash on Ubuntu on Windows thing I'd guess the environment variable just isn't set, and I have no idea what would happen with Cygwin or any other forms of bash on Windows

**Comment by samhutchins** [comment]
Amusingly, `RbConfig::CONFIG['SHELL']` prints `=> "/bin/sh"`

**Comment by donmelton** [comment]
@samhutchins That _is_ funny. :) And obviously not useful. Oh well.

Can you show me the code you used to print out the `cmdcmdline` environment variable?

**Comment by samhutchins** [comment]
In a file called `bar.rb` I have:

```
print "Start"
print ENV["cmdcmdline"]
print "End"
```

**Comment by samhutchins** [comment]
Ignore me, I'm an idiot. I left an echo in my wrapping batch file that confused me. The cmdcmdline thing could well be a red herring

I'll keep digging around, and I'll try not to comment until I actually have something

**Comment by donmelton** [comment]
@samhutchins Thanks! I thought it was probably something like that. :)

OK, then maybe we can write something like this?

```
if ENV['cmdcmdline'] =~ /cmd.exe/i
```

... to test if we're running under `cmd.exe`?

**Comment by donmelton** [comment]
@samhutchins Oh crap, I just saw your previous comment. But Wikipedia says that's a pretty standard variable. See:

https://en.wikipedia.org/wiki/Environment_variable#CMDCMDLINE

**Comment by donmelton** [comment]
@samhutchins Hmmm, but it looks like `cmdcmdline` may be set no matter what shell you're running on Windows. So, yes, it could be a red herring. Damn.

**Comment by samhutchins** [comment]
So when I originally wrote the comment I had a batch file that echo'd `cmdcmdline` before calling Ruby, which is why I thought it worked. Having removed that and more clearly logged the start and end of the Ruby process, calling `puts ENV['cmdcmdline']` doesn't actually print anything, so I guess the variable's not visible to Ruby anyway

You're correct that it's set no matter who calls the file though, but it's set to different things in different cases, which would be enough to differentiate between Powershell and CMD.exe

**Comment by donmelton** [comment]
@samhutchins OK, that's weird since if `cmdcmdline` is set at all in Windows environment, it should be available to Ruby via `ENV['cmdcmdline']`. What happens if you type `ENV['cmdcmdline']` within `irb`?

**Comment by samhutchins** [comment]
Just returns `nil`

`ENV.keys` doesn't list it either, so it's clearly not set in all cases

**Comment by donmelton** [comment]
@samhutchins Dammit. OK, clearly `cmdcmdline` was a dead end. But thanks for testing!

Just out of curiosity, is there anything else in `ENV.keys` that we could look for to test if we're running under `cmd.exe` or Powershell?

**Comment by samhutchins** [comment]
I tried a couple of promising looking things, but there's nothing concrete in there. I'll keep trying things, see how it goes

**Comment by donmelton** [comment]
@samhutchins Thanks! Your help is really appreciated by us all.

**Comment by samhutchins** [comment]
`Process.ppid` returns a pid that's associated with CMD.exe even when I launch `irb` from Powershell

Not sure what else to try 🤔

**Comment by samhutchins** [comment]
I wonder though, is it enough to detect Windows and just return the file path wrapped in `"` in that case? You can't put a `"` in the filename anyway, and it seems to handle `'` just fine. What other characters were messing with it?

Even a path like `"bar '\Rogue One with ' silly characters.mkv"` seems to be fine. 

I've not done a huge amount of testing mind, bit I feel like whatever the user passed to the `detect-crop` command must surely already be escaped for the shell they're using...

**Comment by donmelton** [comment]
@samhutchins Oh, I actually have several files with double quotes in their names. The other characters to special case on are things like question marks, exclamations, colons, etc. We need to escape those somehow depending on the platform.

If we solve this, I want to solve it the right way. Otherwise I can just document the behavior on Windows. After fixing the breakage I caused on Windows with my Z shell change.

**Comment by samhutchins** [comment]
Yay for different rules on different systems. On Windows you can't put a double quote, question mark, or colon in the filename, it won't let you. 

https://support.microsoft.com/en-us/help/905231/information-about-the-characters-that-you-cannot-use-in-site-names,-folder-names,-and-file-names-in-sharepoint

At this point I'd be inclined to fix the issues on zsh without regressions for windows, then document the poor filename escaping. I think the best we can do is simply detect that the program is running on Windows and use some different escaping rules, I just don't think there's a way to detect if the program was launched via CMD or PowerShell.

Also maybe worth a quick read: http://stackoverflow.com/questions/30620876/how-to-properly-escape-filenames-in-windows-cmd-exe

**Comment by donmelton** [comment]
@samhutchins The Sharepoint rules are a bit more restrictive than those for NTFS itself. Go figure. Thanks for the link on escaping. I was doing some research on that too, but, as you said, it doesn't really matter if we can't detect when we're being used from `cmd.exe` or PowerShell.

So, I think at this point we should, as you recommended, 1) fix the regression caused by my Z shell accommodation and 2) document the fact that filename escaping on Windows sucks. Which should be fine since that's how `detect-crop` worked from the beginning and nobody had complained about it until now. :)

**Comment by donmelton** [comment]
@samhutchins BTW, I'll tackle this and merge that dandy patch of yours, #147, when I get back to my development machine tomorrow.

**Comment by donmelton** [comment]
@samhutchins Good news. Using double instead of single quotes around the `lavfi=[drawbox=`... strings works fine in Bash and will _likely_ work in Zsh too. I don't have Zsh installed yet so I can't be sure. I'll tackle that sometime tomorrow. Here's hoping Homebrew can help me out with that.

**Comment by donmelton** [comment]
@samhutchins D'oh! Stupid me. I forgot `zsh` is already part of macOS. :)

Yeah, that works fine. I'll probably make the change later tonight.

**Comment by donmelton** [comment]
@samhutchins I think I've found a way to solve the filename escaping problem! :) At least most of the time anyway. Here's the pseudocode:

```
if arg =~ /([^A-Za-z0-9_\-.,:\/@\n])/
  shell_path = "\"#{arg}\""
else
  shell_path = arg.shellescape
end
```

Basically it will just put double quotes around anything that doesn't need escaping in a Bourne shell. This should be a 80-90% solution when using `cmd.exe` or PowerShell.

What do you think?

**Comment by samhutchins** [comment]
I'd say anything that improves the experience on Windows without breaking stuff elsewhere is a good thing!

I'll modify the gem locally and have a play this evening (BST) to see what kinds of things don't work on Windows, but am I correct in thinking that most of the time the filename will be put in quotes on all platforms?

**Comment by donmelton** [comment]
@samhutchins Yeah, if it doesn't have any spaces, quotes, or any other weird characters, it'll just be enclosed in double quotes.

And thanks! Let me know what you find out.

**Comment by samhutchins** [comment]
So having thought about it, I'm not sure that this really changes much. If the string doesn't need escaping on Unix shells, there's a pretty good chance that it doesn't need escaping on Windows either so wrapping it in quotes doesn't really achieve much, you may as well send back the arg unmodified. I reckon a Platform question is needed:

```
if Gem.win_platform?
    shell_path = "\"#{arg}\""
else
    shell_path = arg.shellescape
end
```
I just don't think there's any way to have common escaping rules between Unix shells and Windows; as soon as you need to escape for Unix shells you break Windows. It does seem to me that full escaping is a PITA on Windows, but wrapping the arg in quotes will deal with 80% of escaping. The Windows file manager won't let me put any of these in a filename: `\/:*?"<>|`.

While escaping rules are a bit different between Powershell and CMD it's definitely not worth the effort to change behaviour on that. Quotes on Windows is better than the current implementation at any rate.

**Comment by donmelton** [comment]
@samhutchins I think Ruby gave all their extra parentheses to Lisp. :)

You make a good point about the usefulness of my pseudocode. I'll have to play with it some more tomorrow.

But I can't add a `Gem.win_platform?` test without breaking escaping when you use Bash or another UNIX-type shell on Windows. Because, I believe, that test will foolishly return `true` in that case.

**Comment by samhutchins** [comment]
@donmelton It does indeed return `true` in Git Bash on Windows. I'd guess the same is true in Cygwin, but I don't fancy installing it to find out I'm afraid.

**Comment by donmelton** [comment]
@samhutchins That's what I figured. Jiminy! And I don't blame you for not wanting to install Cygwin. Try as I might, I could never seem to _uninstall_ the damn thing. :)

And thanks, once again, for all this testing!

**Comment by donmelton** [comment]
@forty2 and @samhutchins OK, the fix for the preview command on Windows is checked in. I'm working on re-writing the "README" now to document the problem with file/path name escaping on Windows.

**Comment by donmelton** [comment]
@samhutchins OK, after this line in the "README" document:

> Just copy and paste the sample commands to preview or transcode.

...I'm going to add this paragraph:

> Please note that path names within the sample commands will not be escaped properly when using `cmd.exe` or PowerShell on Windows.

Does that make sense to you? And is it sufficient?

**Comment by samhutchins** [comment]
@donmelton looks good 👍

**Comment by forty2** [comment]
Hi all,

Sorry for not chiming in sooner; I was traveling and away from my computer.

Thanks for jumping in so quickly and taking this so seriously!  For my purposes, the workaround of just using bash while working on a big batch of videos is good enough, but thanks for fixing the zsh regression (it did seem like this got more annoying recently, but I couldn't put my finger on why...)

The only other thing I would add is you might consider explicitly suggesting `bash.exe` as an alternative in the README comment -- trying bash never occurred to me until you suggested trying PowerShell, so it may not occur to other people either.  Maybe something like:
> Please note that path names within the sample commands will not be escaped properly when using `cmd.exe` or PowerShell on Windows.  If you have Git or Cygwin installed, you can use `bash.exe` instead to work around this issue.

Whatever you think is best; in any case, this bug is sufficiently fixed to solve my problem.  Thanks again!

**Comment by donmelton** [comment]
@forty2 No worries. I've been away from my computer most of the day. They tell me that's a healthy thing to do. :)

That's a good idea, actually. I'd be curious to know what @samhutchins thinks about this and how to word that message.

**Comment by samhutchins** [comment]
I don't like recommending that people use Cygwin, because it's completely unnecessary for this application and it's such a pain to uninstall. But hey, if people already have it and they wanna use it I have no problem with that :-P

I feel that you're the master of wording Don, but I think @forty2 had a pretty good suggestion

**Comment by donmelton** [comment]
@samhutchins Yeah, it was only mentioning "Cygwin" that bothered me. :) Especially after we excised it from the "README" earlier this year and you went the extra mile of writing new documentation.

And I agree, this is a good idea by @forty. So good, in fact, that I delayed the release a day to mull this one over.

OK, I'll come up with some revised wording based on @forty2 's thoughts today, perhaps mentioning Unix shells in general and not Cygwin specifically.

Thanks, again!

**Comment by donmelton** [comment]
@samhutchins and @forty2 OK, here's my reworded text:

> Please note that path names within the sample commands are not escaped properly when using `cmd.exe` or PowerShell on Windows. If you have [Git for Windows](https://git-for-windows.github.io) or another Unix-like environment installed, you can use the [Bash shell](https://en.wikipedia.org/wiki/Bash_(Unix_shell)) — usually named `bash.exe` — to work around this issue.

I'm going with that unless I hear otherwise. :)

**Comment by donmelton** [comment]
@forty2 and @samhutchins All the changes are in version 0.17.3 which I just released. Thanks, again!
